### DISCUSSION TITLE
Batching in CasADi

### DISCUSSION THREAD

**Discussion by Tim-Salzmann** [casadi/casadi#3715] [Q&A] [answered]
Hi,

On the L4CasADi GitHub page, I have recently gotten some questions on better leveraging the efficiency of batched calculations that PyTorch can provide [a].

For example, in our NeRF demo, we used a learned 3D density field as a constraint. Specifically, the density (size [1]) has to be below a threshold at 150 points (size [3] each) along the optimized path. Currently, this is done by iteration [b] over each point and passing each point vector [3] to the external function provided by L4CasADi, which, under the hood, will pass it through the PyTorch model in c++. Similarly, sensitivities will be "requested" from L4CasADi one point at a time. 

Now, PyTorch and thus L4CasADi have the ability and, as a matter of fact, would be much faster to evaluate the function for all points at once in a batch. The input would be [150, 3].
Similarly, the Jacobian could be calculated by batch calculating all 150 Jacobians at once [150, 3, 1].

As far as I understand, however, CasADi naturally expects the Jacobian of the function to be [150, 3, 150, 1]. This would lead to an overhead of calculating 22k extra Jacobians, which will be unused.

My questions, therefore, are:
- Does CasADi have some functionality I am not aware of that leverages such batching efficiencies of an "external" function?
- Or is there a clever way in CasADi to reformulate such a problem that underlying batching efficiencies of an external function could be leveraged?
- If not, do you see a technical path forward to make this feasible?

Appreciate any feedback

Thanks
Tim

[a] https://github.com/Tim-Salzmann/l4casadi/issues/34#issuecomment-2149117104
[b] https://github.com/Tim-Salzmann/l4casadi/blob/4394d56991c712d3e1cc9bd84dde779b28ec3aed/examples/nerf_trajectory_optimization/nerf_trajectory_optimization.py#L77

**Comment by jaeandersson** [comment]
I don't quite understand the issue. Can you explain step-by-step what the dimensions of the external function are and how it's used? Is it 150x3? What do you mean by "[150, 3, 150, 1]"? What derivative information does the external function contain? Is it the Jacobian or directional derivatives?

**Comment by Tim-Salzmann** [comment]
Hi Joel,

Thanks for the quick reply. I think my question was not very well defined, so I tried to understand a bit better how CasADi works for such cases:

```
def cs_model(x):
    return cs.hcat([(cs.sin(x[:, 0]) + cs.cos(x[:, 1]) + cs.tan(x[:, 2])), (cs.cos(x[:, 0]) + cs.cos(x[:, 1]) - cs.tan(x[:, 2]))])

x_sym = cs.MX.sym('x', 10, 3)
y_sym = cs_model(x_sym)
cs_fun = cs.Function('f', [x_sym], [y_sym])
cs_jac_fun = cs.Function('df', [x_sym], [cs.jacobian(cs_fun(x_sym), x_sym)])

inp = np.random.randn(10, 3)
o = cs_jac_fun(inp)
# o -> block diagonal sparse with shape [10*2, 10*3]
```

I now understand that for this example where the function only operates on the column dimension, CasADi will infer this "batch-like" structure and the resulting Jacobian will have block diagonal sparsity.

Now, assuming `cs_fun` is not defined directly in CasADi but is an external function, The external function implements the `jac_fname` [1] symbol. If the external function (generated by L4CasADi) knows about such a structure, can I simply implement the `jac_fname_sparsity_out` symbol to identify this structure back to CasADi? I know from the documentation that this works for `fname_sparsity_out`, but I am not sure if this also applies for the Jacobian and Hessian. It would be great if you could clarify this before I start implementing it.

Going back to the NeRF example in the original post: My goal is to reduce the needed calls to the external L4CasADi (torch) function. 
Going from
```
for k in range(n_eval):
   density = external_nerf_function(points[k])
```
to
```
densities = external_nerf_function(points)

for k in range(n_eval):
   density = densities[k]
   ...
```
which I expect to massively reduce the overhead of context switches from pure c++ to torch and generally improves speed by having one pass per all evaluations/jacobians instead of N.
Currently, with L4CasADi, this is very inefficient as, for example, a dense Jacobian is calculated (mostly zeros) and returned. As I know the structure, L4CasADi could either only calculate the block diagonals and then pad with zeros for the returned dense matrix or (if possible) return the sparsity for derivatives back to CasADi.

> What derivative information does the external function contain? Is it the Jacobian or directional derivatives?

Currently, L4CasADi only implements the Jacobian for the external CasADi functions. Could you elaborate when the directional derivatives would be needed / prefered from the CasADi side?

Appreciate the help!

[1] https://web.casadi.org/docs/#derivatives

**Reply by jaeandersson** [reply, marked as answer]
I think the answer is that it will work as you suggest. If you implement `jac_fname_sparsity_out`, it will be used for the external function. In fact, you should be able to provide `jac_fname_sparsity_out` without even implementing `jac_fname`, in which case it will still be used to estimate the sparsity pattern of the Jacobian (where the actual values will be calculated with say finite differencing).

**Reply by Tim-Salzmann** [reply]
Thanks for confirming. I will give this a try!

Could I ask you to shortly elaborate when the directional derivatives would be needed/preferred from the CasADi side to get an understanding of whether this would be worth being provided by an L4CasADi external function, too?

Thanks
Tim

**Reply by jaeandersson** [reply]
CasADi works with 3 types of derivative information for functions:
 * Forward directional derivatives, in batches of up to 64 directions
 * Adjoint directional derivatives, in batches of up to 64 directions
 * (Full) Jacobians, i.e. the Jacobian of all (differentiable) outputs with respect ot all (differentiable) inputs
 
For symbolic expressions consisting of SX or MX nodes, a heuristic will be used to calculate Jacobians (and Hessians, which are basically symmetric Jacobians). The heuristic is based on graph coloring, using either forward or adjoint directional derivatives (it will test both and choose the more efficient one). More details in the 2018 CasADi paper. That means that forward or adjoint/reverse directional derivatives are propagated forward or backwards through the algorithm. Which means that what will actually be requested from the function will be either forward or adjoint directional derivatives. Now, if a type of derivative information is missing, it can calculate it from another. So if forward/adjoint directional derivatives is missing, it forms the Jacobian and then multiples from the left or right. It may do this too if the number of derivatives is larger than the number of inputs/output elements of the function. Jacobian, if missing, can be calculated from forward or adjoint directional derivatives. If everything is missing, you can still get forward directional derivatives via finite differencing.

To your question - whether it's worth providing directional derivatives.

I think it depends a bit on how the expressions are being used. If all you do with it is to calculate a Jacobian, it might not matter so much. If you need to do adjoint directional derivatives or calculate a Hessian, it might matter a lot.

There might be ways to get even more efficiency though than just providing directional derivatives. In particular, it's possible to have the Jacobian calculation use the external function Jacobian directly and not use any forward or adjoint directional derivatives at all. That would ensure not getting multiple call nodes with the same arguments. It's a bit involved to implement though (it's only been done in some special cases), and there is very little public documentation.

**Reply by Tim-Salzmann** [reply]
Thanks a lot! This makes it much clearer to me!

> There might be ways to get even more efficiency though than just providing directional derivatives. In particular, it's possible to have the Jacobian calculation use the external function Jacobian directly and not use any forward or adjoint directional derivatives at all. That would ensure not getting multiple call nodes with the same arguments. It's a bit involved to implement though (it's only been done in some special cases), and there is very little public documentation.

This indeed sounds like the most efficient solution for the L4CasADi use case. If you have any, could you share some code pointers where this has been done?

Thanks
Tim

**Reply by jaeandersson** [reply]
It works something like this:
 * You extract all the function call nodes using the command "extract" with arguments "lift_shared" set to false and "lift_calls" set to true. See `DaeBuilderInternal::lift`. This gives a new set of variables and equalities defined by: v = v_def(...).
 * You separate the expressions involving function calls from the rest of the variables. That means that your expression gets the form "lifted_expression = lifted_expression_with_calls + lifted_expression_without_calls". In #3708, @jgillis is looking at speeding up the "lifted_expression_without_calls" part by SX expansion by the way so you can look at that issue and code referenced from it.
 * You calculate the Jacobian/Hessian of lifted_expression_without_calls using the normal symbolic routines, possibly after expanding to `SX`. For "lifted_expression_with_calls", you calculate the Jacobians/Hessians for each function and copy them into the right location in the lifted Jacobian/Hessian. For implementation, see `DaeBuilderInternal::jac_vdef_v_from_calls`.
 * You can now calculate the Jacobian/Hessian of lifted_expression as a sum of the Jacobians/Hessians calculated in previous step
 * You eliminate v and v_def from the Jacobian/Hessian expressions. This involves linear systems of the form "I - d(v_def)/dv". Because d(v_def)/dv is a strictly lower triangular matrix, the corresponding linear solve can be done with a simple triangular solve, for which there is an MX atomic.

**Reply by Tim-Salzmann** [reply]
Hi Joel,

I finally found the time to implement the sparsity patterns for batched functions. The good news: the speedup works as expected (factor of 10x for the NeRF example).

However, during implementation one additional question came up:

CasADi seems to expect the following input/output pattern for an external single-input-single-output function:
```
f: [i0] -> [o0]
jac_f: [i0, out_o0] -> [jac_o0_i0]
jac_jac_f: [i0, out_o0, out_jac_o0_i0] -> [jac_jac_o0_i0_i0, jac_jac_o0_i0_out_o0]
```

I am a bit puzzled that there is a second expected output in addition to the hessian, `jac_jac_o0_i0_out_o0` expected by jac_jac_f and could not find any documentation about this. Before this, I guess I was implicitly setting this second output to zero (or random?) by not writing anything back in the buffer for the second output. Now, I am setting it to zero (fully sparse with 0nnz) explicitly.

Could you elaborate on what this second expected output is, how/when it is used by CasADi, and the (unintended) consequences of setting this to zero?

Appreciate the help!
Best
Tim

**Reply by jaeandersson** [reply]
To understand the `jac_jac_o0_i0_out_o0` expression, you need to look at the function signature of `jac_f`. It's the Jacobian of `jac_o0_i0` with respect to `out_o0`. If for example `f` is an implicit function, `out_o0` may be required and thus be non-empty. 

Anyway, are you sure you really want to calculate the "Jacobian of the Jacobian"? That's usually very expensive and not needed by algorithms. Solvers in general need the "Hessian of the Lagrangian", which is a lot cheaper to calculate. In general, probably want to generate:
 * `f`
 * `jac_f`
 * `adj1_f`   (i.e. `f.reverse(1)`)
 * `jac_adj1_f`
